python/ray/train/v2/_internal/callbacks/placement_group_callback.py

```diff
@@ -0,0 +1,126 @@
+import logging
+from typing import TYPE_CHECKING, Optional
+
+import ray
+from ray.exceptions import RayActorError
+from ray.train.v2._internal.execution.callback import (
+    ControllerCallback,
+    WorkerGroupCallback,
+)
+from ray.train.v2._internal.execution.controller.placement_group_cleaner import (
+    PlacementGroupCleaner,
+)
+
+if TYPE_CHECKING:
+    from ray.train.v2._internal.execution.context import TrainRunContext
+    from ray.train.v2._internal.execution.worker_group import WorkerGroup
+
+logger = logging.getLogger(__name__)
+
+
+class PlacementGroupCleanerCallback(ControllerCallback, WorkerGroupCallback):
+    """Callback that manages a PlacementGroupCleaner for the training controller.
+
+    This callback ensures that placement groups are cleaned up even if the controller
+    dies ungracefully.
+    """
+
+    def __init__(self, check_interval_s: float = 1.0):
+        """Initialize the callback.
+
+        Args:
+            check_interval_s: How often (in seconds) the cleaner should check
+                if the controller is still alive.
+        """
+        self._check_interval_s = check_interval_s
+        self._cleaner = None
+        self._controller_actor_id: Optional[str] = None
+
+    def after_controller_start(self, train_run_context: "TrainRunContext"):
+        """Launch the detached PlacementGroupCleaner actor.
+
+        This is called when the controller starts, before the control loop begins.
+        """
+        try:
+            # Launch the cleaner as a detached actor so it survives controller death
+            cleaner_actor_cls = ray.remote(num_cpus=0, max_concurrency=2)(
+                PlacementGroupCleaner
+            )
+            self._cleaner = cleaner_actor_cls.options(
+                lifetime="detached",
+                get_if_exists=False,
+            ).remote(check_interval_s=self._check_interval_s)
+
+            try:
+                core_context = ray.runtime_context.get_runtime_context()
+                self._controller_actor_id = core_context.get_actor_id()
+            except Exception as e:
+                logger.warning(
+                    f"Failed to get controller actor id: {e}. "
+                    "PlacementGroupCleaner will not be able to monitor controller liveness."
+                )
+                return
+
+            if not self._controller_actor_id:
+                logger.warning(
+                    "Controller actor id is unavailable. "
+                    "PG cleaner will not monitor controller liveness."
+                )
+                return
```

Comment:
Should we proactively shut down the cleaner actor here? That way it won't unnecessarily run while the worker group is live.